test/library/draft/DistributedList/DistributedList.chpl

```diff
@@ -0,0 +1,763 @@
+        proc ref set(idx: int, pragma "no auto destroy" in x: eltType): bool {
+            this.lockAll(); defer this.unlockAll();
+
+            if !this.boundsCheck(idx) {
+                _destroy(x);
+                return false;
+            }
+
+            const (locIdx, blockIdx, eltIdx) = indicesFor(idx);
+            this.blockLists[locIdx].set(blockIdx, eltIdx, x);
+            return true;
+        }
+
+        // apply an updater to the value at the given index
+        proc ref update(idx: int, updater) throws {
+            import Reflection;
+
+            this.lockAll(); defer this.unlockAll();
+            assert(this.boundsCheck(idx));
+
+            const (locIdx, blockIdx, eltIdx) = this.indicesFor(idx);
+            // on this.targetLocales[locIdx] {
+                ref slot = this.blockLists[locIdx].getRef(blockIdx, eltIdx);
+
+                if !Reflection.canResolveMethod(updater, "this", idx, slot) then
+                    compilerError('`list.update()` failed to resolve method ' +
+                                updater.type:string + '.this() for arguments (' +
+                                idx.type:string + ', ' + slot.type:string + ')');
+                return updater(idx, slot);
+            // }
+        }
+
+        // return a reference to the element at 'idx'
+        proc ref this(idx: int) ref {
+            assert(this.boundsCheck(idx), "invalid list index: " + idx:string);
+            const (locIdx, blockIdx, eltIdx) = indicesFor(idx);
+            return this.blockLists[locIdx].getRef(blockIdx, eltIdx);
+        }
+
+        proc const isEmpty(): bool {
+            return this.numEntries.read() == 0;
+        }
+
+        inline proc const size {
+            return this.numEntries.read();
+        }
+
+        inline proc const indices {
+            return 0..<this.numEntries.read();
+        }
+
+        // create a block-cyclic distributed array from the distributed list
+        proc toArray(): [] eltType {
+            this.lockAll();
+            const dom = {0..<this.numEntries.read()} dmapped
+                BlockCyclic(startIdx=0, blocksize=this.blockSize, targetLocales=this.targetLocales);
+
+            var a : [dom] this.eltType;
+            coforall (loc, locIdx) in zip(this.targetLocales, this.locDom) do on loc {
+                for (elt, blockIdx, eltIdx) in this.blockLists[locIdx].valsAndIndices() do
+                    a[this.globalIndex(locIdx, blockIdx, eltIdx)] = elt;
+            }
+            this.unlockAll();
+
+            return a;
+        }
+
+        // TODO: make a parallel iterator
+        iter these() {
+            for i in 0..<this.numEntries.read() {
+                const (locIdx, blockIdx, eltIdx) = indicesFor(i);
+                yield this.blockLists[locIdx].getRef(blockIdx, eltIdx);
+            }
+        }
+
+        proc writeThis(fr) throws {
+            fr.write('[');
+            for elt in this.these() {
+                fr.write(elt, ", ");
```

Comment:
Looks like you're not accounting for the end of the list when printing, so the last element has a comma and a space after it